File: examples/basic.c
/* examples/basic.c -- libchroma C API overview */

#include "chroma.h"

#include <stdio.h>
#include <string.h>

/* Helper: unpack a color's channels into a float array and return the count. */
static void print_color(const char *label, chroma_color_t c) {
    float v[4];
    int n = chroma_unpack(c, v);
    printf("%s: (", label);
    for (int i = 0; i < n; i++) printf("%s%.4f", i ? ", " : "", v[i]);
    printf(")\n");
}

int main(void) {
    /* --- hex init + conversion --- */
    printf("--- hex init + conversion ---\n");
    chroma_color_t orange = chroma_init_hex(0xC86432);
    printf("input:   #%06X\n", chroma_unpack_hex(orange));
    print_color("srgb", orange);
    print_color("oklch", chroma_convert(orange, CHROMA_OKLCH));
    print_color("hsl", chroma_convert(orange, CHROMA_HSL));

    /* --- gamut mapping --- */
    printf("\n--- gamut mapping ---\n");
    chroma_color_t p3_green = chroma_init(CHROMA_DISPLAY_P3, (float[]){0.0f, 1.0f, 0.0f});
    print_color("input", p3_green);
    printf("in srgb gamut: %s\n", chroma_is_in_gamut(p3_green, CHROMA_SRGB) ? "yes" : "no");
    chroma_color_t mapped = chroma_gamut_map(p3_green, CHROMA_SRGB);
    print_color("mapped", mapped);
    uint8_t r, g, b;
    chroma_unpack_srgb8(mapped, &r, &g, &b);
    printf("srgb8:   (%d, %d, %d)\n", r, g, b);
    printf("hex:     #%06X\n", chroma_unpack_hex(mapped));

    /* --- alpha --- */
    printf("\n--- alpha ---\n");
    chroma_alpha_color_t semi = chroma_init_hexa(0xC8643280);
    float alpha_vals[4];
    float alpha;
    chroma_unpack_alpha(semi, alpha_vals, &alpha);
    printf("input:   #C8643280\n");
    printf("rgb:     (%.4f, %.4f, %.4f)\n", alpha_vals[0], alpha_vals[1], alpha_vals[2]);
    printf("alpha:   %.2f\n", alpha);

    /* --- null hue --- */
    printf("\n--- null hue ---\n");
    chroma_color_t grey = chroma_init_hex(0x808080);
    chroma_color_t grey_hsl = chroma_convert(grey, CHROMA_HSL);
    float hsl[4];
    chroma_unpack(grey_hsl, hsl);
    printf("input:   #808080 (grey)\n");
    printf("hsl:     (h=%s, s=%.4f, l=%.4f)\n",
        chroma_hue_is_null(hsl[0]) ? "null" : "???", hsl[1], hsl[2]);

    /* --- interactive prompt --- */
    printf("\n--- interactive ---\n");
    char buf[64];

    int count = chroma_space_count();
    printf("space (");
    for (int i = 0; i < count; i++) {
        if (i > 0) printf(", ");
        printf("%s", chroma_space_name(i));
    }
    printf("): ");
    if (!fgets(buf, sizeof(buf), stdin)) return 1;
    buf[strcspn(buf, "\n")] = '\0';

    int space_val = chroma_space_from_name(buf);
    if (space_val < 0) { fprintf(stderr, "unknown space: %s\n", buf); return 1; }
    chroma_space_t space = (chroma_space_t)space_val;

    int n = chroma_field_count(space);
    float vals[4];
    for (int i = 0; i < n; i++) {
        printf("value %d: ", i + 1);
        if (scanf("%f", &vals[i]) != 1) return 1;
    }

    chroma_color_t clr = chroma_init(space, vals);

    printf("\nhex:     #%06X\n", chroma_unpack_hex(clr));
    print_color("srgb", chroma_convert(clr, CHROMA_SRGB));
    print_color("oklch", chroma_convert(clr, CHROMA_OKLCH));

    return 0;
}

File: examples/basic.zig
// examples/basic.zig -- libchroma Zig API overview

const std = @import("std");
const chroma = @import("libchroma");

var line_storage: [128]u8 = undefined;

fn readLine(reader: *std.Io.Reader) ?[]const u8 {
    const line = reader.takeDelimiterInclusive('\n') catch return null;
    const without_delim = if (line.len > 0 and line[line.len - 1] == '\n') line[0 .. line.len - 1] else line;
    const trimmed = std.mem.trim(u8, without_delim, &.{ ' ', '\r', '\t', '"', '\'' });
    if (trimmed.len > line_storage.len) return null;
    @memcpy(line_storage[0..trimmed.len], trimmed);
    return line_storage[0..trimmed.len];
}

pub fn main(init: std.process.Init) !void {
    var out_buf: [4096]u8 = undefined;
    var w = std.Io.File.stdout().writer(init.io, &out_buf);
    const out = &w.interface;

    // --- static demos ---

    // hex init + conversion
    try out.writeAll("--- hex init + conversion ---\n");
    const orange = chroma.Srgb(f32).initFromHex(0xC86432);
    const oklch = chroma.convert(orange, chroma.Oklch(f32));
    try out.print("input:  #C86432\n", .{});
    try out.print("oklch:  L={d:.2} C={d:.2} H={?d:.2}\n\n", .{ oklch.l, oklch.c, oklch.h });

    // gamut mapping
    try out.writeAll("--- gamut mapping ---\n");
    const p3_green = chroma.DisplayP3(f32).init(0, 1, 0);
    const mapped = chroma.gamut.gamutMap(p3_green, chroma.Srgb(f32));
    try out.print("input:  display_p3(0.00, 1.00, 0.00)\n", .{});
    try out.print("mapped: srgb({d:.2}, {d:.2}, {d:.2})\n\n", .{ mapped.r, mapped.g, mapped.b });

    // alpha
    try out.writeAll("--- alpha ---\n");
    const semi = chroma.Alpha(chroma.Srgb(f32)).init(orange, 0.5);
    const semi_oklch = chroma.convert(semi, chroma.Alpha(chroma.Oklch(f32)));
    try out.print("input:  srgb #C86432 with alpha 0.50\n", .{});
    try out.print("alpha preserved after conversion: {d:.2}\n", .{semi_oklch.alpha});

    try out.flush();

    // --- interactive prompt ---

    try out.writeAll("\n--- runtime ---\n");

    var inbuf: [256]u8 = undefined;
    var reader = std.Io.File.stdin().reader(init.io, &inbuf);
    const in = &reader.interface;

    const space_names = comptime blk: {
        const fields = @typeInfo(chroma.Space).@"enum".fields;
        var names: [fields.len][]const u8 = undefined;
        for (fields, 0..) |f, i| names[i] = f.name;
        break :blk names;
    };
    try out.writeAll("space (");
    for (space_names, 0..) |name, i| {
        if (i > 0) try out.writeAll(", ");
        try out.writeAll(name);
    }
    try out.writeAll("): ");
    try out.flush();

    const space_name = readLine(in) orelse return;
    const space = std.meta.stringToEnum(chroma.Space, space_name) orelse {
        try out.print("unknown space: {s}\n", .{space_name});
        try out.flush();
        return;
    };

    const field_names = chroma.color.fieldNames(space);
    var vals: [4]f32 = undefined;
    for (field_names, 0..) |name, i| {
        try out.print("{s}: ", .{name});
        try out.flush();
        const val_str = readLine(in) orelse return;
        vals[i] = std.fmt.parseFloat(f32, val_str) catch {
            try out.print("invalid number: {s}\n", .{val_str});
            try out.flush();
            return;
        };
    }

    const input_color = chroma.color.initFromSlice(space, vals[0..field_names.len]) catch unreachable;

    const srgb = chroma.color.convert(input_color, .srgb);
    const oklch_out = chroma.color.convert(input_color, .oklch);

    try out.writeAll("\n");
    try out.print("{s}: {f}\n", .{ @tagName(space), chroma.color.formatter(input_color, .default) });
    try out.print("srgb: {f}\n", .{chroma.color.formatter(srgb, .default)});
    try out.print("oklch: {f}\n", .{chroma.color.formatter(oklch_out, .default)});
    try out.flush();
}
